server/src/main/java/org/cloudfoundry/identity/uaa/zone/MultitenantVmidentityClientDetailsService.java

```diff
@@ -0,0 +1,335 @@
+        } catch (NoSuchOIDCClientException e) {
+            logger.error("No client found with id = " + clientId, e);
+            throw new NoSuchClientException("No client found with id = " + clientId);
+        } catch (Exception e) {
+            logger.error("Unable to remove client details with id = " + clientId, e);
+            throw new VmidentityDataAccessException("Unable to remvoe client details");
+        }
+    }
+
+    @Override
+    public List<ClientDetails> listClientDetails() {
+        try {
+            String tenant = VmidentityUtils.getTenantName(idmClient);
+            Collection<OIDCClient> clients = idmClient.getOIDCClients(tenant);
+
+            List<ClientDetails> details = new ArrayList<ClientDetails>(clients.size());
+            for (OIDCClient client : clients) {
+                details.add(mapOIDCClient(client, tenant, idmClient, passwordEncoder));
+            }
+
+            return details;
+        } catch (Exception e) {
+            logger.error("Unable to list client details for identity zone: " + IdentityZoneHolder.get().getName(), e);
+            throw new VmidentityDataAccessException("Unable to list client details");
+        }
+    }
+
+    @Override
+    public ClientDetails loadClientByClientId(String clientId) throws ClientRegistrationException {
+        ClientDetails details;
+
+        try {
+            String tenant = VmidentityUtils.getTenantName(idmClient);
+            details = mapOIDCClient(idmClient.getOIDCClient(tenant, clientId), tenant, idmClient, passwordEncoder);
+        } catch (NoSuchOIDCClientException e) {
+            logger.error("Unable to load client with id: " + clientId, e);
+            throw new NoSuchClientException("No client with requested id: " + clientId);
+        } catch (Exception e) {
+            logger.error("An error occurred loading client with id: " + clientId, e);
+            throw new VmidentityDataAccessException("Unable to load client with id: " + clientId);
+        }
+
+        return details;
+    }
+
+    @Override
+    public int deleteByIdentityZone(String zoneId) {
+        try {
+            String tenant = VmidentityUtils.getTenantName(zoneId, idmClient);
+            Collection<OIDCClient> clients = idmClient.getOIDCClients(tenant);
+
+            for (OIDCClient client : clients) {
+                idmClient.deleteOIDCClient(tenant, client.getClientId());
+            }
+
+            return clients.size();
+        } catch (Exception e) {
+            logger.error("Unable to delete clients for identity zone: " + zoneId, e);
+            throw new VmidentityDataAccessException("Unable to delete clients for identity zone: " + zoneId);
+        }
+    }
+
+    @Override
+    public int deleteByOrigin(String origin, String zoneId) {
+        return 0;
+    }
+
+    @Override
+    public Log getLogger() {
+        return logger;
+    }
+
+    @Override
+    public int getTotalCount() {
```

Comment:
are we sure this gives a count for ALL tenants? did you check uaa impl? it looks like it might be a bug from their end if this is a multi-tenant and all other methods are using ZoneHolder ...